query_class_config.hh

```diff
@@ -23,19 +23,40 @@
 
 #include <cinttypes>
 
+namespace ser {
+
+template <typename T>
+class serializer;
+
+};
+
+
 namespace query {
 
 struct max_result_size {
     uint64_t soft_limit;
     uint64_t hard_limit;
+private:
+    uint64_t page_size = 0;
+public:
 
     max_result_size() = delete;
     explicit max_result_size(uint64_t max_size) : soft_limit(max_size), hard_limit(max_size) { }
     explicit max_result_size(uint64_t soft_limit, uint64_t hard_limit) : soft_limit(soft_limit), hard_limit(hard_limit) { }
+    max_result_size(uint64_t soft_limit, uint64_t hard_limit, uint64_t page_size)
+            : soft_limit(soft_limit)
+            , hard_limit(hard_limit)
+            , page_size(page_size)
+    { }
+    uint64_t get_page_size() const {
+        return page_size == 0 ? hard_limit : page_size;
```

Comment:
On old nodes, on regular (non-reversed) queries, the page size was actually limited by default to 1MB - which wasn't (IIUC) hard_limit but rather some other constant. Doesn't the new code break this?

Reply:
The limit is sent by the coordinator to replicas for many years now. It is a constant, yes, but it is set by the coordinator in a per-query manner. This doesn't change that.

Reply:
But it's not clear *how* it doesn't change this. On **older** nodes page_size=0 so this returns hard_limit, not the 1MB page size. Maybe somehow this code only effects reversed queries and not regular queries?
I am not claiming there is something wrong in your code - just saying that for readers a few months for now (and to some degree, for me even today) it is hard to understand how all of these parameters fit together, and a comment explaining everything might have been helpful.

Reply:
The current (old) code tries to squeeze 3 parameters into 2 variables, by observing that at any given time, maximum 2 are needed. The two variables are the two existing member of `max_query_size`: `hard_limit` and `soft_limit`. There are two cases:
* Normal query, only needs page size: `hard_limit == soft_limit == $page_size`;
* Unbounded query (reverse or unpaged), doesn't need page size, only needs the soft/hard limits: `soft_limit = $soft_limit`, `hard_limit = `$hard_limit`.

The problem this PR tries to solve is that with the new native reverse queries support, reverse queries potentially need all of these. So we need a 3rd variable. The games with `page_size == 0` that you see here is for supporting old nodes. These will send just the first two variables, set to `$hard_limit` and `$soft_limit` respectively, but no separate page size. So we use `$hard_limit` as page size in this case. When the coordinator is a new one and sends a page size too (and its not 0) we use that value.

I agree this should be described somewhere, the best place would be just above `struct max_result_size`.